src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -749,23 +942,55 @@ int16_t LoRaWANNode::downlink(uint8_t* data, size_t* len) {
       remLen -= processedLen;
       foptsPtr += processedLen;
     }
+
+    // if FOptsLen for the next uplink is larger than can be piggybacked onto an uplink, send separate uplink
+    if(this->commandsUp.len > 15) {
+      uint8_t foptsNum = this->commandsUp.numCommands;
+      size_t foptsBufSize = this->commandsUp.len;
+      uint8_t foptsBuff[foptsBufSize];
+      size_t idx = 0;
+      for(size_t i = 0; i < foptsNum; i++) {
+        LoRaWANMacCommand_t cmd = { .cid = 0, .len = 0, .payload = { 0 }, .repeat = 0, };
+        popMacCommand(&cmd, &this->commandsUp, i);
+        if(cmd.cid == 0) {
+          break;
+        }
+        foptsBuff[idx] = cmd.cid;
+        for(size_t i = 1; i < cmd.len; i++) {
+          foptsBuff[idx + i] = cmd.payload[i];
+        }
+        idx += cmd.len + 1;
+      }
+      RADIOLIB_DEBUG_PRINTLN("Uplink MAC payload (%d commands):", foptsNum);
+      RADIOLIB_DEBUG_HEXDUMP(foptsBuff, foptsBufSize);
+
+      isMACPayload = true;
+      this->uplink(foptsBuff, foptsBufSize, RADIOLIB_LORAWAN_FPORT_MAC_COMMAND);
+    }
+
+    // write the MAC command queue to nvm for next uplink
+    uint8_t queueBuff[sizeof(LoRaWANMacCommandQueue_t)];
+    memcpy(&queueBuff, &this->commandsUp, sizeof(LoRaWANMacCommandQueue_t));
+    mod->hal->writePersistentStorage(mod->hal->getPersistentAddr(RADIOLIB_PERSISTENT_PARAM_LORAWAN_FOPTS_ID), queueBuff, sizeof(LoRaWANMacCommandQueue_t));
```

Comment:
Is it necessary to save this in NVM?

Reply:
If the device is going to sleep after a downlink, the MAC responses need to be stored for the next uplink, otherwise they would be lost. Especially on ESP32 boards where the RAM is discarded through deepsleep (and powercycle).

Reply:
In a later revision @StevenCellist may want to check best practise - sending an uplink to ack some MAC commands before sleeping if the sleep is going to be a while may be a good idea - particularly if there are more needed and/or the queue took priority over a user level downlink that may have some import.

TL;DR: Keep a short cycle going until all the messages have been processed.